op-node/rollup/derive/channel_in_reader.go

```diff
@@ -53,80 +39,50 @@ func (cr *ChannelInReader) Progress() Progress {
 	return cr.progress
 }
 
+// TODO: Take full channel for better logging
 func (cr *ChannelInReader) WriteChannel(data []byte) {
 	if cr.progress.Closed {
 		panic("write channel while closed")
 	}
-	cr.data = data
-	cr.ready = false
-}
-
-// ReadBatch returns a decoded rollup batch, or an error:
-//   - io.EOF, if the ChannelInReader source needs more data, to be provided with WriteChannel()/
-//   - any other error (e.g. invalid compression or batch data):
-//     the caller should ChannelInReader.NextChannel() before continuing reading the next batch.
-func (cr *ChannelInReader) ReadBatch(dest *BatchData) error {
-	// The channel reader may not be initialized yet,
-	// and initializing involves reading (zlib header data), so we do that now.
-	if !cr.ready {
-		if cr.data == nil {
-			return io.EOF
-		}
-		if cr.r == nil {
-			cr.r = bytes.NewReader(cr.data)
-		} else {
-			cr.r.Reset(cr.data)
-		}
-		if cr.readZlib == nil {
-			// creating a new zlib reader involves resetting it, which reads data, which may error
-			zr, err := zlib.NewReader(cr.r)
-			if err != nil {
-				return err
-			}
-			cr.readZlib = zr.(zlibReader)
-		} else {
-			err := cr.readZlib.Reset(cr.r, nil)
-			if err != nil {
-				return err
-			}
-		}
-		if cr.readRLP == nil {
-			cr.readRLP = rlp.NewStream(cr.readZlib, 10_000_000)
-		} else {
-			cr.readRLP.Reset(cr.readZlib, 10_000_000)
-		}
-		cr.ready = true
+	if f, err := BatchReader(bytes.NewBuffer(data), cr.progress.Origin); err == nil {
+		cr.nextBatchFn = f
+	} else {
+		cr.log.Error("Error creating batch reader from channel data", "err", err)
 	}
-	return cr.readRLP.Decode(dest)
 }
 
 // NextChannel forces the next read to continue with the next channel,
 // resetting any decoding/decompression state to a fresh start.
 func (cr *ChannelInReader) NextChannel() {
-	cr.ready = false
-	cr.data = nil
+	cr.nextBatchFn = nil
 }
 
 func (cr *ChannelInReader) Step(ctx context.Context, outer Progress) error {
 	if changed, err := cr.progress.Update(outer); err != nil || changed {
 		return err
 	}
 
-	var batch BatchData
-	if err := cr.ReadBatch(&batch); err == io.EOF {
+	if cr.nextBatchFn == nil {
+		return io.EOF
+	}
+
+	// TODO: can batch be non nil while err == io.EOF
+	// This depends on the behavior of rlp.Stream
```

Comment:
Good comment, checking this now

Reply:
one of the todo's that I have is to switch over to using a custom sentinel error `EOD` which has the semantics that if `EOD` is returned, no data is returned.

Reply:
`io.EOF` is not returned early. It does return when it attempts to read incomplete data, which can be the kind of rlp entry upon reading the first byte, or anything of the value content later. But if the underlying reader returns an early EOF after returning sufficient data, then it's zeroed, based on `readFull` decs (only reader, except for the byte-by-byte reader (which would transform an `io.EOF` upon reading a single byte to a `io.ErrUnexpectedEOF`)
```
			// Readers are allowed to give EOF even though the read succeeded.
			// In such cases, we discard the EOF, like io.ReadFull() does.
```

Reply:
Edit: on the top-level of the rlp decoding stack it does return `EOF` rather than `ErrUnexpectedEOF` when reading of the first element fails. So it will hit the `EOF` case correctly here.